### PR TITLE
Count Update Sync & Shot Caller Interface

### PR DESCRIPTION
**Pull request by geenen124**
Also:
- Renamed XMLHelper to ProjectManager and added Director Timeline parsing.
- Refactored Sockets to Improve Testability.

### CODE REVIEW

**Comment by Markvanderuit** [comment]
Looks good, second member can merge.

File: client_app/js/shotCaller.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// eslint-disable-next-line no-undef
+document.addEventListener("DOMContentLoaded", () => {
+    const socket = io("/shotCallers"); // eslint-disable-line no-undef
+
+    socket.on("next count", (countData) => {
```
Comment (bart-de-jonge) [inline client_app/js/shotCaller.js]: I don't think you should use spaces here, maybe use underscores or camelcase

File: client_app/js/shotCaller.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        document.getElementsByTagName("shot-caller-display")[0]
+            .setAttribute("current-count", countData.newCount);
+    });
+
+    socket.on("current director shot", (data) => {
```
Comment (bart-de-jonge) [inline client_app/js/shotCaller.js]: no spaces

File: client_app/js/shotCaller.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        document.getElementsByTagName("shot-caller-display")[0]
+            .setAttribute("current-shot", JSON.stringify(data.currentShot));
+    });
+
+    socket.on("next director shot", (data) => {
```
Comment (bart-de-jonge) [inline client_app/js/shotCaller.js]: no spaces

File: routes/sockets/ShotCallerSocket.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    const nextShot = this.findNextShot(directorTimeline);
+                    socket.emit("next director shot", {
+                        nextShot,
+                    });
+                });
+            });
+        });
+    }
+
+    sendNextCount(newCount) {
```
Comment (bart-de-jonge) [inline routes/sockets/ShotCallerSocket.js:66]: missing comment

File: routes/sockets/ShotCallerSocket.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            setTimeout(waitForXML, 10);
+        } else {
+            const directorTimeline = projectManager.data.directorTimeline;
+            callback(directorTimeline);
+        }
+    }
+    waitForXML();
+};
+
+class ShotCallerSocket {
```
Comment (bart-de-jonge) [inline routes/sockets/ShotCallerSocket.js:21]: missing comment

File: routes/sockets/ShotCallerSocket.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
+/**
+ * socket.io handlers for shot callers.
+ * Exposes methods to enable broadcasting of shot events.
+ */
+import log4js from "log4js";
+import ProjectManager from "../../objects/ProjectManager";
+
+const logger = log4js.getLogger();
+
+const getDirectorTimeline = callback => {
```
Comment (bart-de-jonge) [inline routes/sockets/ShotCallerSocket.js:11]: missing comment

File: routes/sockets/CameraOperatorSocket.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            logger.info("New Camera Operator Connection");
+            // Initialize With Current Server Count
+            this.sendNextCount(this.currentCount);
+            socket.on("advance count", () => {
+                this.advanceCountCallBack();
+            });
+        });
+    }
+
+    sendNextCount(newCount) {
```
Comment (bart-de-jonge) [inline routes/sockets/CameraOperatorSocket.js:35]: missing comment

File: routes/sockets/CameraOperatorSocket.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+/**
+ * socket.io handlers for camera operators.
+ * Exposes methods to enable broadcasting of shot events.
+ */
+import log4js from "log4js";
+
+const logger = log4js.getLogger();
+
+class CameraOperatorSocket {
```
Comment (bart-de-jonge) [inline routes/sockets/CameraOperatorSocket.js:12]: missing comment

File: routes/sockets/CameraOperatorSocket.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    constructor(io, currentCount, advanceCountCallBack) {
```
Comment (bart-de-jonge) [inline routes/sockets/CameraOperatorSocket.js:19]: missing comment

File: routes/router.js [hunk 10]
```diff
@@ -5,10 +5,12 @@ import index from "./index";
 import timeline from "./timeline";
 import uploadScp from "./upload-scp";
 import camera from "./camera";
+import shotCaller from "./shot-caller";
 
 module.exports.addRoutes = (app) => {
```
Comment (bart-de-jonge) [inline routes/router.js:8]: missing comme nt

File: objects/ProjectManager.js [hunk 11]
```diff
@@ -1,15 +1,35 @@
...
+let projectManagerInstance = null;
+
 /*
- * Class for storing a CameraTimeline
+ * Class for storing a Project
  */
-class XMLHelper {
+class ProjectManager {
+
+    constructor() {
```
Comment (bart-de-jonge) [inline objects/ProjectManager.js:19]: missing comment

File: objects/DirectorShot.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    constructor(name, description, beginCount, endCount, frontShotPadding, endShotPadding) {
+        this.name = name;
+        this.description = description;
+        this.beginCount = beginCount;
+        this.endCount = endCount;
+        this.frontShotPadding = frontShotPadding;
+        this.endShotPadding = endShotPadding;
+    }
+
+    static fromXML(XMLObject) {
```
Comment (bart-de-jonge) [inline objects/DirectorShot.js:24]: missing comment

File: objects/DirectorShot.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Class for storing a Director Shot.
+ */
+class DirectorShot {
+    constructor(name, description, beginCount, endCount, frontShotPadding, endShotPadding) {
```
Comment (bart-de-jonge) [inline objects/DirectorShot.js:14]: missing comment

File: routes/sockets/sockets.js [hunk 14]
```diff
@@ -4,45 +4,53 @@
...
+import ShotCallerSocket from "./ShotCallerSocket";
+import ProjectManager from "../../objects/ProjectManager.js";
 
 const logger = log4js.getLogger();
 
+// Get timelines from xml
+const getTimelines = function getTimelines(callback) {
+    // Dummyfile Todo: replace with dyn0amic
+    const projectManager = new ProjectManager();
+    function waitForXML() {
```
Comment (bart-de-jonge) [inline routes/sockets/sockets.js]: missing comment

Comment (bart-de-jonge) [inline routes/sockets/sockets.js]: Maybe extract this function to ProjectManager and work with a callback

**Comment by bart-de-jonge** [comment]
Looks good

**Pull request merged**